### ISSUE TITLE
Collect and store SCP messages in vault client

### ISSUE THREAD

**Issue by ebma** [pendulum-chain/spacewalk#85]
### Collecting externalized messages
- Connect to the Stellar overlay network
- Listen to the gossiped `StellarMessage`s 
    - Filter them by their contained `pledge`; we are only interested in`SCPStatementExternalize` messages and group them by their `slotIndex`
    - To make it easier to later find the relevant set of externalized messages given a specific transaction, store them as described below

### Storing externalized messages
- Choose the `txSetHash` contained in one of the externalized messages (should be the same for all externalized messages in the same slot)
- Fetch the transaction set from the overlay network (more details see below)
- For all the transactions in the transaction set, build the transaction hash and store them in a map (`Map<transactionHash, slotIndex>`) with their transaction hash as the key, and the slotIndex (of the externalized messages they belong to) as the value. The same map is used for all transactions sets so that it eventually contains the transaction hashes of all the transactions we know (unless pruned after some timeframe). 
- Store all externalized messages with the same `slotIndex` together with the related transaction set in a map with the `slotIndex` as the key. Something like `Map<slotIndex, {externalizedMessages[], transactionSet[]}>`

These steps have to be repeated for every 'slot'. 
The `Map<transactionHash, slotIndex>` is just for convenience to make the lookup more efficient.

### Fetching a transaction set from the overlay network
- Use a `GET_TX_SET` [StellarMessage](https://stellar-sdk.readthedocs.io/en/latest/api.html#stellarmessage) to receive the transaction set from the overlay network
- It is not yet clear how we would do this in code

### Building the transaction hash
- The transactions we receive from the overlay network do not have a transaction hash, so we have to derive it ourselves
- The transaction hash is generated by hashing the `signatureBase` with sha256 (see [here](https://github.com/stellar/js-stellar-base/blob/master/src/hashing.js) and [here](https://github.com/stellar/js-stellar-base/blob/437e823d4529f2ff38ce6e8c7ca01fac3365ce21/src/transaction_base.js#L198))
- The derivation of the `signatureBase` of a 'normal' transaction is shown [here](https://github.com/stellar/js-stellar-base/blob/437e823d4529f2ff38ce6e8c7ca01fac3365ce21/src/transaction.js#L259) and the one of a fee-bump transaction is shown [here](https://github.com/stellar/js-stellar-base/blob/437e823d4529f2ff38ce6e8c7ca01fac3365ce21/src/fee_bump_transaction.js#L81). But I think it is safe to ignore the fee-bump transaction because we would not process these kinds of transactions in the oracle agent

**Comment by b-yap** [comment]
The **`GET_TX_SET`** needs the `txsethash`. I'm not exactly sure where/how to get it. 
```
pub struct ScpStatementExternalize {
    pub commit: ScpBallot,
    pub n_h: Uint32,
    pub commit_quorum_set_hash: Hash,
}
```
> Choose the txSetHash contained in one of the externalized messages. 

It's not the `commit_quorum_set_hash`... or is it?

From the types I've seen, it's in the `StellarValue` [struct](https://stellar-sdk.readthedocs.io/en/latest/api.html#stellarvalue)
I don't know where I should be looking at.

**Comment by ebma** [comment]
Hey @b-yap. The `txsethash` is in the `StellarValue` which itself is in the `ScpBallot`. It is not the `commit_quorum_set_hash`; that is something else.

To get a better understanding, you can find a graphical representation of the structure [here](https://www.notion.so/satoshipay/Stellar-Oracle-Structure-v2-d5feb20060444f93822746bc8ceb9e9e#978b0c025624416990c60e8f2af45f33).

**Comment by b-yap** [comment]
@ebma 
For `Map<slotIndex, {externalizedMessages[], transactionSet[]}>`, Are we really saving the entire set? The transactionSet could have a size of 271, 314, ... sometimes more than 500... 

There's only 1 externalized message per slot, isn't it? Based on the structure of `ScpBallot`?

And also do we have a marker, a limit on when to stop storing the externalized messages?

**Comment by ebma** [comment]
Sorry, I should have made this clear in this ticket. (I only pointed it out in the [notion](https://www.notion.so/satoshipay/Stellar-Oracle-Structure-v2-d5feb20060444f93822746bc8ceb9e9e#978b0c025624416990c60e8f2af45f33) document)

> To verify that a transaction happened on the Stellar network, the bridge pallets need: 
> - The `externalizedMessages[]` of a full quorum set of tier 1 validator nodes for a specific slot
>     - To be precise, the `externalizedMessages[]` are `SCPEnvelope`s that contain `SCPStatementExternalize` messages in them (see [[here](https://www.notion.so/Stellar-Oracle-Structure-v2-d5feb20060444f93822746bc8ceb9e9e)](https://www.notion.so/Stellar-Oracle-Structure-v2-d5feb20060444f93822746bc8ceb9e9e)).
>         - This is necessary because the `SCPEnvelope`s contain a signature that makes it possible to **verify** **the** **whole** **statement**, whereas the `SCPStatementExternalize` also contains a signature, but this only allows for **verification of the StellarValue** (i.e. among others, the `txSetHash`) that was agreed on but not the `commitQuorumSetHash` which is needed to verify the quorum


So to be precise the `externalizedMessages[]` that we store actually have to be `ScpEnvelopes` that contain the externalized statement.

And I didn't expect the transaction set to be so large. But I think we actually need the entire set because the spacewalk pallet has to check if the to-be-verified transaction is included in the tx set and then re-build the `txsethash` from the entire set of transactions to verify it. 

I can't tell yet how much storage this would take up, but I'd say you could implement something that entries/slots older than ~1 hour can be dropped from the map. We can later adjust that parameter. Or did you mean something else with 'marker'?

**Comment by ebma** [comment]
Although it is not necessary to store _all_ ScpEnvelopes, we agreed on storing all of them for a certain timeframe for the time being. 
We should implement a smart mechanism that drops messages from storage that are not required anymore. But this has to consider that messages and transactions may be delayed due to latency and that the parachain could enter an error state, where fulfilling issue/redeem requests is impossible. 

The period we use for storing the messages is still to be defined. We could measure it empirically and decide depending on how much space it takes.

**Comment by b-yap** [comment]
I'm saving old data in file. And with Gonza's idea, we can use a script to delete those files in a certain time.
The saving of old data would be based on the size of the map: an agreed MAX size; like `MAX_SLOTS_PER_FILE` of 500.

Currently separating both the ScpEnvelope from the TransactionSet to different maps.  
I'm having problems with storing them together if we consider the time factor, since the `TxSet` messages are coming too late and the map is getting bigger while waiting for it. 

ex.
slot 1, 2, 3, 4, 5 ScpEnvelopes have been filled; but we still don't have the txset for slot 3. Then we got slot 6, 7, 8, 9 ScpEnvelopes in succession.
Note: for every slot, approx ~ 20 scpenvelopes are received ; that's 100 messages we've received, and still no txset for slot 3. It also happened that we actually get txsets for slots 6 and 7. 

I tried to send `GetTxSet` again for slot 3, but it seems to take so much time.  
Separating them would be easier; if all scpenvelopes are received, we can just write it to file (whether or not we get the txset.) 
But again, it's a todo. A dirty code in [Stellar-Experiments](https://github.com/pendulum-chain/stellar-experiments/blob/85-collect-and-store-scp-messages-in-vault-client/src/bin/example.rs) and slowly transferring to spacewalk.
